### DISCUSSION TITLE
Controller error: Function components cannot be given refs

### DISCUSSION THREAD

**Discussion by noxd3v** [react-hook-form/react-hook-form#7864] [Q&A] [answered]
### Version Number

7.19.1

### Codesandbox/Expo snack

https://codesandbox.io/s/polished-shadow-x81106

### Steps to reproduce

Can you please tell me why am I getting this error in the console?

### Expected behaviour

Ref should have been forwarded with props

### What browsers are you seeing the problem on?

Chrome

### Relevant log output

```shell
Warning: Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?

Check the render method of `App`.
    at AutosuggestInput (https://x81106.csb.app/src/AutoSuggestInput.js:89:28)
    at div
    at App (https://x81106.csb.app/src/App.js:36:41)
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by Moshyfawn** [comment]
#3857

**Reply by noxd3v** [reply]
> It shows same error even if I put `ref` on rendered input without spreading the `field` prop

```javascript
   <Controller
      render={({ field: { value, ref, ...otherProps } }) => (
        <Wrapper>
          <Input value={value || ""} ref={ref} {...otherProps} />
          <Results
            input={value}
            compare={compareField}
            options={populateWith}
            render={({ searchResults, setSearchResults }) => {
              return searchResults?.length > 0 ? (
                <ResultsWrapper>
                  {searchResults.map((result) => (
                    <ResultItem
                      key={result[valueField]}
                      onClick={() => {
                        onSelect(result[valueField], result[compareField]);
                        setSearchResults([]);
                      }}
                    >
                      {result[labelField]}
                    </ResultItem>
                  ))}
                </ResultsWrapper>
              ) : null;
            }}
          />
        </Wrapper>
      )}
      {...otherProps}
    />
```

**Reply by Moshyfawn** [reply]
You're passing a `ref` from the parent component into the `Controller` by spreading register and then `otherProps` into the `Controller`.
You shouldn't mix `register` and `Controller`. See the comment below from @bluebill1049 about the same issue

**Comment by bluebill1049** [comment, marked as answer]
hey, @webjoyable this is a setup issue, from my observation:

1. your component already has Controller, so you don't need to register again.
![Screen Shot 2022-02-19 at 9 09 18 pm](https://user-images.githubusercontent.com/10513364/154796534-b5024bad-dc77-478c-9fb3-726b06620cee.png)

https://react-hook-form.com/api/usecontroller/controller
![Screen Shot 2022-02-19 at 9 10 29 pm](https://user-images.githubusercontent.com/10513364/154796563-97442850-0d3f-46cc-8288-44e2a3073e87.png)

2. why are you getting that react warning because you didn't forward the ref.  https://reactjs.org/docs/forwarding-refs.html

**Reply by noxd3v** [reply]
> 1. your component already has Controller, so you don't need to register again.

`field` doesn't have `register` in it, I just received `onChange, onBlur, ref, name, value`

> 2. why are you getting that react warning because you didn't forward the ref.  https://reactjs.org/docs/forwarding-refs.html

You are right about this one. I tried it before, but I was forwarding `ref` to `Controller` component instead of rendered `input`.

Thank you.

Working example:

```javascript
const AutosuggestInput = React.forwardRef(
  (
    {
      populateWith,
      loading,
      labelField,
      compareField,
      valueField,
      onSelect,
      ...otherProps
    },
    ref
  ) => {
    return loading ? (
      <Spinner />
    ) : (
      <Controller
        render={({ field: { value, ...otherFieldProps } }) => (
          <Wrapper>
            <Input {...otherFieldProps} value={value || ""} ref={ref} />
            <Results
              input={value}
              compare={compareField}
              options={populateWith}
              render={({ searchResults, setSearchResults }) => {
                return searchResults?.length > 0 ? (
                  <ResultsWrapper>
                    {searchResults.map((result) => (
                      <ResultItem
                        key={result[valueField]}
                        onClick={() => {
                          onSelect(result[valueField], result[compareField]);
                          setSearchResults([]);
                        }}
                      >
                        {result[labelField]}
                      </ResultItem>
                    ))}
                  </ResultsWrapper>
                ) : null;
              }}
            />
          </Wrapper>
        )}
        {...otherProps}
      />
    );
  }
);
```
